spec/system/admin/bulk_product_update_spec.rb

```diff
@@ -527,13 +527,14 @@
     s2 = create(:supplier_enterprise)
     p1 = FactoryBot.create(:simple_product, name: "product1", supplier: s1)
     p2 = FactoryBot.create(:simple_product, name: "product2", supplier: s2)
-    login_as_admin
 
-    visit spree.admin_products_path
+    login_as_admin_and_visit spree.admin_products_path
```

Comment:
Would it be better practice to declare these variables and under a `context` block with `let`, and add a `before` to the routing? I'm happy to try it out in a separate PR and compare results with out new bash routine :sparkles:

Reply:
> Would it be better practice to declare these variables and under a context block with let

Yes, I think so.

> and add a before to the routing?

With routing you mean the `visit` statement? I often found it hard to add more test cases with different conditions to the same block if you have the page load already done in the `before` block. Sometimes you want to add something like an additional product or revoke permissions before the page is loaded.

So generally, I think:

* Setup context like data in the database or date and time in `let` and `before`.
* Perform the action and expectations in the `it` block.

One common pattern is to have three lines or three blocks in an `it` block:

```rb
it "works with :foo" do
  input = :foo # this is optional and can happen with `let` if re-used in other blocks
  result = do_the_work_with(input)
  expect(result)
end
```

I like to use `let` when it makes sense to group the data like `let(:apple)`, `let(:banana)` and so on. But if you have a single variable which is only used in one `it` block then I would declare it there. Keep things close together so that you don't have to scroll far. A `let` at the beginning of the file which is used 300 lines further down is often annoying to look up and therefore it's difficult to spot mistakes.

Reply:
>  I often found it hard to add more test cases with different conditions to the same block if you have the page load already done in the before block.

Indeed, I've found no reference mentioning that having the `visit` statement in a `before` block reduces flakiness. So, yeah, happy to keep it inside the `it` of course. In this particular spec though `login_as_admin_and_visit spree.admin_products_path` repeats throughout within several `it` blocks, so I think it could be useful to extract it into a `before` block, following DRY. But no strong opinion. 

> I like to use let when it makes sense to group the data like let(:apple), let(:banana) and so on.

Agree not to use `let` and simply declare the variable with equal `=`, if we only do it once :+1: 

> A let at the beginning of the file which is used 300 lines further down is often annoying to look up and therefore it's difficult to spot mistakes.

However, I did not mean to set it in the beginning of the file, but rather close to the `it` block, on its respective `context`, as suggested here:

https://www.betterspecs.org/#let

Totally agree that it should not compromise readability - on the contrary, we should aim to improve it.

Do you consider this one a valid reference have been other references been followed so far?  I usually look up these:
https://www.betterspecs.org/
https://rspec.rubystyle.guide/

Are there others?
Pinging @jibees here as well.

Reply:
> Do you consider this one a valid reference have been other references been followed so far?

* Rubocop style guide: yes. I just read through half of that document and mostly agree. The only exception in my opinion is around using `context` and `specify` instead of `describe` and `it`. As a beginner, I found it confusing what all the different things are. So with several options doing the same thing, I prefer to stick to one. But that's just my opinion and Rubocop is a widely acknowledged style guide. Happy to go with that.
* Better Specs: this has some overlap with Rubocop, some other good points and some more opinionated comments. It's not as extensive and not updated as much. It's a good read but also not that important.

I don't have any other references. Over the years I've read many blog posts which addressed one or several issues and some of these stuck because I liked them and others I forgot.

I'm not sure if we should create our own wiki with additions or try to contribute to https://rspec.rubystyle.guide/.

I try to suggest only things with a reason, for example when I encountered a bug which was only possible because the spec was unclear. I'm just another opinionated human though.